community/values/src/main/java/org/neo4j/values/Value.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.values;
+
+public abstract class Value
+{
+    @Override
+    public abstract boolean equals( Object other );
+
+    @Override
+    public abstract int hashCode();
+
+    public abstract boolean equals( Value other );
+
+    public abstract boolean equals( byte[] x );
+
+    public abstract boolean equals( short[] x );
+
+    public abstract boolean equals( int[] x );
+
+    public abstract boolean equals( long[] x );
+
+    public abstract boolean equals( float[] x );
+
+    public abstract boolean equals( double[] x );
+
+    public abstract boolean equals( boolean x );
+
+    public abstract boolean equals( boolean[] x );
+
+    public abstract boolean equals( char x );
+
+    public abstract boolean equals( String x );
+
+    public abstract boolean equals( char[] x );
+
+    public abstract boolean equals( String[] x );
```

Comment:
These `equals`-methods are not going to have very interesting implementations for most sub-classes. In particular for "extension" value types - the ones we are going to implement in the Cypher runtime, i.e. `NodeReference`, `ValueMap`, `Point`, et.c.

How about giving them a default implementation that returns `false`?

Perhaps we should also consider "sealing" them, so that such "extension" value types _cannot_ override these methods. We could do that through a pattern like this:

```
public final boolean equals( String[] that ) {
    return eq( that );
}
boolean eq( String[] that ) {
    return false; // default implementation
}
```